### ISSUE TITLE
Docker 적용

### ISSUE THREAD

**Issue by wjrmffldrhrl** [kyu9341/TeamHash#82]
지금 웹 서버 배포를 하고있는 AWS는 몇가지 문제점이 존재합니다.
1. 인스턴스 종료시 모든 환경 초기화
2. 윈도우 환경에서 테스트시 동작하는 기능이 AWS 위에서 동작하지 않음 (ex: progress bar)
3. 프리티어기간이 곧 만료됨

이에 가상 컨테이너를 구축할 수 있는 Docker를 도입하려고 합니다.

현재 최대한 비슷하게 구성해 둔 이미지를 [Docker Hub](https://hub.docker.com/repository/docker/wjrmffldrhrl/teamhash)에 올려놨습니다.

테스트중이며 이후에 Dockerfile과 Docker Compose로 빌드 및 배포 자동화를 구현할 것입니다.
이외에도 Spring Boot에서 Docker 적용을 편리하게 해주는 [플러그인](https://mvnrepository.com/artifact/com.spotify/docker-client)도 존재한다고 해서 알아보는 중입니다.

지금은 MariaDB와 Spring Boot가 한 컨테이너 안에 있지만 나중에는 두개를 분리시켜 관리하는 방법도 생각하고있습니다.

**Comment by ytw9699** [comment]
궁금한게 있어요~

2번에서 윈도우 환경에서 테스트시 동작하는 기능이 AWS 위에서 동작하지 않는다고 하셨는데

프로젝트 aws 환경을 찾아보니 https://github.com/kyu9341/TeamHash/issues/52

 os가 Ubuntu Server 20.04 LTS라고 적혀있는데, Ubuntu가 맞나요~?

**Comment by wjrmffldrhrl** [comment]
네, AWS OS는 Ubuntu Server 20.04 LTS로 ubuntu가 맞습니다. 

윈도우에서 로컬에 배포하여 접속했을 때 잘 작동하던 프로그래스바가 AWS상에서 배포하여 접속했을 때 동작을 하지 않는 것을 확인했습니다.

**Comment by ytw9699** [comment]
그렇다면 그냥 제 생각인데~

프로그래스바가 윈도우에서만 동작하는거라면,

ec2의 os를 우분투가 아닌 윈도우로 한다면 프로그래스바가 aws 위에서도 동작 하지않을까요~?

**Comment by wjrmffldrhrl** [comment]
거기까지는 생각을 못했는데 그렇게 한다면 동작할 것 같습니다.

한번 테스트를 해봐야겠어요

그래도 컨테이너의 이점을 경험해보고싶어서 Docker 환경 구성은 계속 해볼 생각입니다.

좋은 아이디어 감사합니다!

**Comment by ytw9699** [comment]
네 그리고 하나 더 궁금한게, 계속 도커로 환경 구성을 하신다면 도커에서도 os 환경 선택을 해야하지 않나요? 우분투인지, 윈도우인지요~

**Comment by wjrmffldrhrl** [comment]
지금은 우분투를 기준으로 진행중이지만 그부분은 좀더 고민을 해봐야 될 것 같습니다.

배포는 리눅스에서 해야지~ 라는 생각으로 우분투만 생각하고있었는데 윈도우라는 선택지도 있네요

웹 서버를 배포할 때 어떤 OS가 더 많은 이점이 있을까요?

**Comment by ytw9699** [comment]
저도 배포는 우분투로 하는데 처음에 우분투를 선택한 이유가, 윈도우보다 가볍다는 이점이 있다고 해서 선택했었거든요~

그런데 계속 프로젝트 진행하다보니 os환경이 달라서 윈도우에서 돌리던 환경을 우분투에 맞게 바꿔줘야 하는..

번거로움이 많더라구요. 지금도.. 그것때매 맞춰줘야 하는게 많아 힘든데..

그래서 제가 이 이슈를 봤을때, 도커를 도입하시는 이유에 1,3번은 공감이 되요. 그런데 2번에서는 좀 의문이 들어요~

저 프로그래스바를 윈도우가 아닌 우분투의 환경설정에 맞게 바꿔준다해도 동작이 안되시는거죠~?

**Comment by wjrmffldrhrl** [comment]
아직 우분투 환경에서 프로그래스바에 대해 환경설정을 바꾼다거나 조치사항을 적용한 것은 없고 AWS 상에서 동작이 되지 않습니다.

문제를 접하고 어느 부분을 손대야할지 감이 잡히질 않아서 방치를 해둔면이 없지 않아 있습니다.

환경이 달라서 생기는 문제를 경험한 것이 이번 프로젝트에서의 프로그래스바가 처음이라 많이 당황했었습니다.

**Comment by ytw9699** [comment]
아 그러셨군요 네 알겠습니다~ 근데 제 생각이 아닐지도 몰라요 뇌피셜이라ㅎㅎ

하여튼 늦게까지 열심히시네요 좋은 밤 되시길^^

**Comment by jongfeel** [comment]
Docker 세팅 후 AWS 배포의 경우 철저히 실행 방법 및 상태 체크 관련된 작업이라고 생각하고 해야 합니다.

- 사전 준비
- 배포 방법 명령 혹은 세팅 파일
- 현재 실행된 상태 확인 후 추가로 체크해서 해야 할 부분
- 이 모든 것에 대한 매뉴얼

이건 다른 어떤 것 보다 매뉴얼 작업이 매우 중요합니다.
따라서 매뉴얼 작성 후 서로 cross-check 한다는 생각으로 체크해 주면 좋습니다.
